Support for Olimex ESP32-C3-DevKit-Lipo

simonidelh opened the pull request:
Prot of gravitymon to olimex ESP32C3 board with charger on board. 

**Tested:**
- Gyro
- Battery-Level

Added docs about wiring and the solder briges to get the battery level

mp-se commented:
Looks like you got all the places changed, i just have one challenge that needs to be adressed (which i need to do) and the defines for the target platform is also used in my espframework so for that to be configured correctly then the ESP32C3 needs to be defined as well...

but if i refactor the code and environment config to differ between CPU and BOARD then that would work. The board definition should be used to define the PIN mapping only. The board is already defined for the environment so that needs to be picked up for the compiler to include the right code.

mp-se commented:
I just noticed one issue and that is you are using pin 0 and 1 for two purposes which is not good.

#define PIN_SCL 1
#define PIN_SDA 0
#define PIN_DS 5
#define PIN_VOLT 3
#define PIN_CFG1 0
#define PIN_CFG2 1

You need to change either the config or the I2C pins. It would also be good if you use the pin definitions in the pins_arduino.h for that board.

simonidelh commented:
Hey @mp-se ,
thanks for the feedback. I've fixed that and found some issues with the onboard LED, which I now have fixed. Also, I've added some docs. No Schematics, since I've got no tooling installed, so I've written it down.

simonidelh commented:
> Looks like you got all the places changed, i just have one challenge that needs to be adressed (which i need to do) and the defines for the target platform is also used in my espframework so for that to be configured correctly then the ESP32C3 needs to be defined as well...
> 
> but if i refactor the code and environment config to differ between CPU and BOARD then that would work. The board definition should be used to define the PIN mapping only. The board is already defined for the environment so that needs to be picked up for the compiler to include the right code.

Well, the issue with the LED showed me, that the olimex board is not defined, so I've had to use a patch....

mp-se commented:
Hi did some refactoring in the code to support custom boards since this is not the only request for that. 

* The board ID will now be used for the PIN configuration in the main.hpp file. A script will inject a define with the board ID so that can be used to map the pins correctly. See: scripts/board.py
* Firmware script will now also copy the file and name it custom-board.bin unless its one of the targets that I can test. This way its visible which ones are custom. So adding a custom board should not require any changes to that script. 

One thing remains to be adressed and that is the OTA support since that is in the espframework, I need to change that so the filename can be set from gravitymon. 

Please have a look at the changes and give me some feedback.

simonidelh commented:
> Hi did some refactoring in the code to support custom boards since this is not the only request for that.
> 
>     * The board ID will now be used for the PIN configuration in the main.hpp file. A script will inject a define with the board ID so that can be used to map the pins correctly. See: scripts/board.py
> 
>     * Firmware script will now also copy the file and name it custom-board.bin unless its one of the targets that I can test. This way its visible which ones are custom. So adding a custom board should not require any changes to that script.
> 
> 
> One thing remains to be adressed and that is the OTA support since that is in the espframework, I need to change that so the filename can be set from gravitymon. I will also look into how to add this to the documentation.
> 
> Please have a look at the changes and give me some feedback.

Hmmmm... But my board is not defiened by PIO, so I have to patch a simmelar one to fit my needs. This BoardID, I can't unse to set my pins, since another board with the same BoardID would destory the config, right?

simonidelh commented:
If you want better management of the boards: What about a dependend include? If newBoard, then include `boardname/pins_arduino.h` and `boardname/pindev.h`
or something simmilar.
Then your can still use your boarddefining mechanism and I would have a standarized way, to patch boards, not included in PIO

mp-se commented:
> If you want better management of the boards: What about a dependend include? If newBoard, then include `boardname/pins_arduino.h` and `boardname/pindev.h` or something simmilar. Then your can still use your boarddefining mechanism and I would have a standarized way, to patch boards, not included in PIO

That is already handled by platformio behind the scenes. Patching an existing board is probably not a good idea, better to add the new board definition and includes to platformio before the build. There is another PR that has the same challenge. 

It might be that upgrading to Arduino 3.x would solve part of this if the board is defined there but the issue for me is that Arduino 3 increase the size of the bin so it will no longer fit in the partition table. This is the main reason for not taking that step. 

Let me have a look at this and see if I can suggest an approach to patch arduino before build so we can continue to leverage the functionallity in platformio. I dont really want to start creating a lot of workaround for different boards that will be hard to maintain. I did patch the arduino before they fixed the LED setting for the s3 mini board so I have something i can start with.

simonidelh commented:
I will have a look into board definitions and get back with a new draft. Your totally right, to enforce the managable way.

I'm thinkin about a `custom-board.json` and `pin_arduino.h` in the repo, which gets patched. Should be the correct way, if I'm not mistaken.
This way all custom stuff is managable in one location (due to the fact, that the definition never should change)

mp-se commented:
I've added a script to patch platformio with new board defintions to the dev branch. Just copy the files to the correct location in the patch directory, script is run before a build

simonidelh commented:
actually, the patching is way more easy than I expected... `.json`-files in the root at `boards/someboard.json` works flawlessly.
I've used your `patch_arduino.py` to copy the `someboard.pins.h` to `../variants/someboard/arduino_pins.h`.

Works perfectly... 
Any toughts about this change?

simonidelh commented:
Switched to variant-directory to completly get away from patching the buildsystem with some custom libs...

mp-se commented:
> actually, the patching is way more easy than I expected... `.json`-files in the root at `boards/someboard.json` works flawlessly. I've used your `patch_arduino.py` to copy the `someboard.pins.h` to `../variants/someboard/arduino_pins.h`.
> 
> Works perfectly... Any toughts about this change?
> 
> And probably, I should merge into `dev` instead of `master`?

Patching platformio and arduino is quite simple, just add the files in the correct location. Here is a good guide / example on what needs to be done (this is related to the other PR that is open) but it might give some pointers on what to consider when adding a new board. I have tested the build process with these files. https://sigmdel.ca/michel/ha/esp8266/super_mini_esp32c3_en.html

I can change the target, but yes I try to keep the development outside the master branch so that reflects the latest release only.

mp-se commented on src/gyro.cpp:
Why have you changed this ? Does not really add anything except that it takes one second longer to initialize the gyro.

mp-se commented on src/main.hpp:
This look good!

mp-se commented on src_docs/source/hardware.rst:
Can you move this to the new custom.rst file instead ? The ++ below the headline needs to be the same length as the headline for the build to work.

mp-se commented on platformio.ini:
I've set a naming standard for the targets so can you adjust the environment name to:

[ev:gravity-olimex_esp32_c3_devkit_lipo]

For the build to work properly the define ESP32C3 needs to be added this is also used in the underlying espframework, the ESP32C3_OLIMEX should not be needed since the define for the board should be used in main.hpp

I would assume the ARDUINO_USB_CDC_ON_BOOT would be needed if this board has an USBC port or the serial output would probably not work properly. Or have you tested that ?

simonidelh replied:
thanks for the refactor. I will check the build. Serial via USB was working, will get back to you after testing

mp-se commented on boards/olimex_esp32_c3_devkit_lipo/pins_arduino.h:
If you move this folder to the platforms/framework-arduino... then my new script should copy the files to the correct location. No need to have the boards directory or add other scripts.

simonidelh replied:
The `boards` directory is the way to go for custom boards. That's why I used this aproach. No custom scripting needed. If you are opposed to this thing, I will build it with your scripts - just let me know.
https://docs.platformio.org/en/latest/projectconf/sections/platformio/options/directory/boards_dir.html

mp-se replied:
If there is a standard way then that would be better, let me check your approach and i will get back

mp-se replied:
I stand corrected, your approach to handle the board definitions are way better than my approach. I will revert my changes in favor of your suggestion.

mp-se commented on script/patch_arduino.py:
My patch_platformio script would do the same as this if the files are placed in the right directory

simonidelh replied:
ah, i forgot to delete this file. With `/boards` and the `variant_dir`, there is no need to patch. I will implement whatever suits you better.

mp-se commented:
I think the board definition file for platformio is missing in the PR

mp-se commented:
Thanks. now it compiles. I have also added it to the github action for the automated builds. I have also added a change for the OTA support but I still need to verify that it works correctly. I will add it to the webflasher as well and once that is done I would appreciate if you can test it.

simonidelh commented:
> Thanks. now it compiles. I have also added it to the github action for the automated builds. I have also added a change for the OTA support but I still need to verify that it works correctly. I will add it to the webflasher as well and once that is done I would appreciate if you can test it.

Sure, im up for testing. Just let me know when

mp-se commented:
It was blocked by a statement in .gitignore, should be there now. I had to copy it manually since i already had started the merge locally.

mp-se commented:
I have added a test of the flasher here https://gravitymon.com/flasher/test.html

simonidelh commented:
> I have added a test of the flasher here https://gravitymon.com/flasher/test.html

Will do, but at the moment the device is swimming in a IPA. I will get back to you after the fermentation

mp-se commented:
Its no rush, hope the ipa will be good. Thats my favorite style!

simonidelh commented:
Webflasing is also working. Thank you :)

mp-se commented:
Thanks, i will publish it then
